src/MANYBODY/pair_eam_he.cpp

```diff
@@ -0,0 +1,238 @@
+{
+  he_flag = 1;
+}
+
+void PairEAMHE::compute(int eflag, int vflag)
+{
+  int i,j,ii,jj,m,inum,jnum,itype,jtype;
+  double xtmp,ytmp,ztmp,delx,dely,delz,evdwl,fpair;
+  double rsq,r,p,rhoip,rhojp,z2,z2p,recip,phip,psip,phi;
+  double *coeff;
+  int *ilist,*jlist,*numneigh,**firstneigh;
+
+  evdwl = 0.0;
+  ev_init(eflag,vflag);
+
+  // grow energy and fp arrays if necessary
+  // need to be atom->nmax in length
+
+  if (atom->nmax > nmax) {
+    memory->destroy(rho);
+    memory->destroy(fp);
+    memory->destroy(numforce);
+    nmax = atom->nmax;
+    memory->create(rho,nmax,"pair:rho");
+    memory->create(fp,nmax,"pair:fp");
+    memory->create(numforce,nmax,"pair:numforce");
+  }
+
+  double **x = atom->x;
+  double **f = atom->f;
+  int *type = atom->type;
+  int nlocal = atom->nlocal;
+  int nall = nlocal + atom->nghost;
+  int newton_pair = force->newton_pair;
+
+  inum = list->inum;
+  ilist = list->ilist;
+  numneigh = list->numneigh;
+  firstneigh = list->firstneigh;
+
+  // zero out density
+
+  if (newton_pair) {
+    for (i = 0; i < nall; i++) rho[i] = 0.0;
+  } else for (i = 0; i < nlocal; i++) rho[i] = 0.0;
+
+  // rho = density at each atom
+  // loop over neighbors of my atoms
+
+  for (ii = 0; ii < inum; ii++) {
+    i = ilist[ii];
+    xtmp = x[i][0];
+    ytmp = x[i][1];
+    ztmp = x[i][2];
+    itype = type[i];
+    jlist = firstneigh[i];
+    jnum = numneigh[i];
+
+    for (jj = 0; jj < jnum; jj++) {
+      j = jlist[jj];
+      j &= NEIGHMASK;
+
+      delx = xtmp - x[j][0];
+      dely = ytmp - x[j][1];
+      delz = ztmp - x[j][2];
+      rsq = delx*delx + dely*dely + delz*delz;
+
+      if (rsq < cutforcesq) {
+        jtype = type[j];
+        p = sqrt(rsq)*rdr + 1.0;
+        m = static_cast<int> (p);
+        m = MIN(m,nr-1);
+        p -= m;
+        p = MIN(p,1.0);
+        coeff = rhor_spline[type2rhor[jtype][itype]][m];
+        rho[i] += ((coeff[3]*p + coeff[4])*p + coeff[5])*p + coeff[6];
+        if (newton_pair || j < nlocal) {
+          coeff = rhor_spline[type2rhor[itype][jtype]][m];
+          rho[j] += ((coeff[3]*p + coeff[4])*p + coeff[5])*p + coeff[6];
+        }
+      }
+    }
+  }
+
+  // communicate and sum densities
+
+  if (newton_pair) comm->reverse_comm_pair(this);
+
+  // fp = derivative of embedding energy at each atom
+  // phi = embedding energy at each atom
+  // if rho > rhomax or rho < rhomin (i.e., table is exceeded),
+  //   add linear term to conserve energy
+
+  for (ii = 0; ii < inum; ii++) {
+    i = ilist[ii];
+    p = (rho[i]-rhomin)*rdrho + 1.0;
+    m = static_cast<int> (p);
+    if (m < 2) {
+      m = 2;
+    } else if (m > nrho-1) {
+      m = nrho-1;
+    }
+    p -= m;
+    if (p < -1.0) {
+      p = -1.0;
+    } else if (p > 1.0) {
+      p = 1.0;
+    }
+    coeff = frho_spline[type2frho[type[i]]][m];
+    fp[i] = (coeff[0]*p + coeff[1])*p + coeff[2];
+    if (eflag) {
+      phi = ((coeff[3]*p + coeff[4])*p + coeff[5])*p + coeff[6];
+      if (rho[i] < rhomin) {
```

Comment:
It seems that the only difference in compute() relative to pair_eam.cpp is right here.  Could this be handled in another way, without duplicating code?

Reply:
I already removed some duplicated code from the original version sent to me (which was a modified version of pair_eam_fs.cpp with the modified compute() method added), but didn't see a simple way to do this without impacting performance and complicating the entire "tree" of derived classes. This seems to be a bit of a niche application, so I stopped at this and did not consider to generalize the change like we just did in tersoff (which would be a similar if not larger effort with the added complication that the contributing author is not as comfortable with git and github and thus asked me to help getting this contribution included). A major difference to the tersoff case is that this modification also requires a modified potential file (and file reader). The entire difference between the two compute methods is given below. If you feel that could/should be merged, you are probably best off contacting the contributing author (at SNL) directly via e-mail.
```
@@ -85,21 +84,33 @@
 
   // fp = derivative of embedding energy at each atom
   // phi = embedding energy at each atom
-  // if rho > rhomax (e.g. due to close approach of two atoms),
-  //   will exceed table, so add linear term to conserve energy
+  // if rho > rhomax or rho < rhomin (i.e., table is exceeded),
+  //   add linear term to conserve energy
 
   for (ii = 0; ii < inum; ii++) {
     i = ilist[ii];
-    p = rho[i]*rdrho + 1.0;
+    p = (rho[i]-rhomin)*rdrho + 1.0;
     m = static_cast<int> (p);
-    m = MAX(1,MIN(m,nrho-1));
+    if (m < 2) {
+      m = 2;
+    } else if (m > nrho-1) {
+      m = nrho-1;
+    }
     p -= m;
-    p = MIN(p,1.0);
+    if (p < -1.0) {
+      p = -1.0;
+    } else if (p > 1.0) {
+      p = 1.0;
+    }
     coeff = frho_spline[type2frho[type[i]]][m];
     fp[i] = (coeff[0]*p + coeff[1])*p + coeff[2];
     if (eflag) {
       phi = ((coeff[3]*p + coeff[4])*p + coeff[5])*p + coeff[6];
-      if (rho[i] > rhomax) phi += fp[i] * (rho[i]-rhomax);
+      if (rho[i] < rhomin) {
+        phi += fp[i] * (rho[i]-rhomin);
+      } else if (rho[i] > rhomax) {
+        phi += fp[i] * (rho[i]-rhomax);
+      }
       phi *= scale[type[i]][type[i]];
       if (eflag_global) eng_vdwl += phi;
       if (eflag_atom) eatom[i] += phi;
```